Fully working IME - WASM, ios and Android

danwalmsley opened the pull request:
Combined effort: @Gillibald @kekekeks @danwalmsley 

- win32 preedit support.
- android pre edit
- ios preedit


- refactors and fixes to ime api, with working android and ios keyboard.
- wasm - sync ime context with browser.


Remaining todo:

- [x] Return key setting
- [x] Focus issue on android.
- [x] pressing button when keyboard is showing should trigger button command as well as close keyboard.
- [x] Accepts return obeyed

Gillibald committed:
Rendering part of PreeditText

Gillibald committed:
Imm32 preedit

Gillibald committed:
Cleanup

Gillibald committed:
Merge remote-tracking branch 'upstream/master' into feature/ImePreeditText

Gillibald committed:
Merge remote-tracking branch 'upstream/master' into feature/ImePreeditText

Gillibald committed:
Android preedit

Gillibald committed:
Adjust remaining InputMethodImpl

Gillibald committed:
Merge branch 'master' into feature/ImePreeditText

Commit:
Merge branch 'master' into feature/ImePreeditText

Gillibald committed:
Fix iOS build

Gillibald committed:
Merge remote-tracking branch 'origin/feature/ImePreeditText' into feature/ImePreeditText

Gillibald committed:
Disable special preedit highlight

Commit:
add simple mobile sandbox apps.

Commit:
Merge branch 'feature/ImePreeditText' of github.com:Gillibald/Avalonia into feature/ImePreeditText

Commit:
Merge branch 'master' into feature/ImePreeditText

Commit:
Merge remote-tracking branch 'origin/master' into feature/ImePreeditText

Commit:
whitespace.

Commit:
add implementation of advanced ios ime api.

Commit:
remove some assets.

Commit:
add implementation of returnkeytype for ios.

Commit:
add hack implementation of surrounding text.

Commit:
further ios ime implementation.

Commit:
make implementation consistent with ios sample.

https://developer.apple.com/library/archive/samplecode/SimpleTextInput/Introduction/Intro.html#//apple_ref/doc/uid/DTS40010633

Commit:
use seperate responder class

Commit:
add fix for window becoming keywindow.

Commit:
fix responder issue.

kekekeks committed:
Working on iOS IME

Commit:
implement force press spacebar on ios.

Commit:
make android compatible with new ime interfaces.

Gillibald committed:
Android - Initial attempt to fix focus management and implement SurroundingText

Commit:
Merge remote-tracking branch 'origin/master' into ios-ime-responder

# Conflicts:
#	src/Web/Avalonia.Web.Blazor/AvaloniaView.razor.cs
#	src/Windows/Avalonia.Win32/Input/Imm32InputMethod.cs

maxkatz6 commented on src/Avalonia.Base/Input/TextInput/TextInputReturnKeyType.cs:
```diff
@@ -0,0 +1,17 @@
+    Route,
+    Search,
+    Send,
+    Yahoo,
```
Do we need something as specific as Yahoo or Google here? Wouldn't it be enough to use Search instead?

kekekeks replied:
Where did we get those values from? Looks looks something OS-specific

danwalmsley replied:
https://developer.apple.com/documentation/uikit/uireturnkeytype

danwalmsley replied:
used a small subset that is cross platform.

Commit:
initial implementation of ime interop for web

Commit:
Merge remote-tracking branch 'origin/master' into ios-ime-responder

Gillibald committed:
Android IME - Try suggestion handling

Gillibald committed:
Merge branch 'ios-ime-responder' of https://github.com/AvaloniaUI/Avalonia into ios-ime-responder

Gillibald committed:
Properly handle mutiple suggestions

Commit:
add callbacks for wasm ime, ready to sync with me.

Commit:
ios implement return key type

Commit:
implement return key and accepts return, and closing keyboard on done return keys.

Commit:
limited set of return key types for cross platform.

Commit:
basic implementation of android return key type.

Commit:
wasm: seperate inputhelper and focushelper to prevent multiple events.

Gillibald committed:
Handle preedit

Gillibald committed:
Merge branch 'ios-ime-responder' of https://github.com/AvaloniaUI/Avalonia into ios-ime-responder

Gillibald committed:
Move all text input to the InputHelperInterop

Commit:
support net 7 on wasm backend.

Commit:
Merge branch 'ios-ime-responder' of github.com:AvaloniaUI/Avalonia into ios-ime-responder

Commit:
add support for positioning ime suggestions.

Commit:
Merge remote-tracking branch 'origin/master' into ios-ime-responder

Gillibald committed:
Reset surroundingText on focus change

Reset surroundingText on focus change

Gillibald committed:
Reset InputElement.value on input event

Commit:
add caret helper javascript.

Commit:
Merge branch 'ios-ime-responder' of github.com:AvaloniaUI/Avalonia into ios-ime-responder

Gillibald committed:
Attempt to sync caret and surroundingText

Gillibald committed:
Merge branch 'ios-ime-responder' of https://github.com/AvaloniaUI/Avalonia into ios-ime-responder

Commit:
add caret helper functions.

Commit:
wasm: keep ime and surrounding text in sync and aligned carets (managed / native)

Commit:
dispatch the CursorRectangleEvent invoke so recipients will see textbox / textpresenter / ime apis in a consistent state.

Commit:
dont set textbox width.

Gillibald committed:
Merge

Gillibald committed:
Remove TextBefore, TextAfter

Set surroundingText relative to current line
Raise SurroundingTextChanged on CursorRectangleChanged

Commit:
ensure caretRectChange events propogate

Commit:
make input box size to the surrounding text and position correctly.

Commit:
make native input box invisible.

Commit:
Merge remote-tracking branch 'origin/master' into ios-ime-responder

Commit:
remove mobile sandbox apps from build.

Gillibald committed:
Properly raise surroundingText changed

Commit:
Merge remote-tracking branch 'origin/master' into ios-ime-responder

Gillibald committed:
Fix GetTextLineText

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024016-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

Commit:
tweak wasm inputbox so not visible but takes part in a11y tree for ios compat.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024020-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

Gillibald committed:
Android - switch to full editor input connection

Gillibald committed:
Merge branch 'ios-ime-responder' of https://github.com/AvaloniaUI/Avalonia into ios-ime-responder

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024027-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

danwalmsley commented:
we decided to merge this as only a few remaining issues with WASM implementation, and that will be better dealt with in PR for new wasm backend.
